### ISSUE TITLE
The training code

### ISSUE THREAD

**Issue by ZHUYMGeo** [ZJier/DKDMN#4]
Good work! but you did not provide the training code to perform the contrastive learning, it seems that the reader needs to write the related codes.

**Comment by ZJier** [comment]
整体训练和测试部分的代码，你可以参考：https://github.com/nshaud/DeepHyperX 或 https://github.com/xiachangxue/DeepHyperX，你也可以很容易将这个工作嵌入进去。
对比学习部分的训练代码并不是独立的，它可以在上述两个链接中直接嵌入集成。下面是一个示例，你可以进行相关尝试：
首先，关于损失函数部分的代码如下：
#===============================================================
class ContrastiveLoss(nn.Module):
    def __init__(self, batch_size, device='cuda', temperature=0.5):
        super().__init__()
        self.batch_size = batch_size
        self.register_buffer("temperature", torch.tensor(temperature).to(device)) # 超参数 温度
        self.register_buffer("negatives_mask", (~torch.eye(batch_size * 2, batch_size * 2, dtype=bool).to(device)).float()) # 主对角线为0，其余位置全为1的mask矩阵
    def forward(self, emb_i, emb_j): # emb_i, emb_j 是来自同一图像的两种不同的预处理方法得到
        z_i = F.normalize(emb_i, dim=1) # (bs, dim) ---> (bs, dim)
        z_j = F.normalize(emb_j, dim=1) # (bs, dim) ---> (bs, dim)
        representations = torch.cat([z_i, z_j], dim=0) # repre: (2*bs, dim)
        similarity_matrix = F.cosine_similarity(representations.unsqueeze(1), representations.unsqueeze(0), dim=2) # simi_mat: (2*bs, 2*bs)
        sim_ij = torch.diag(similarity_matrix, self.batch_size) # bs
        sim_ji = torch.diag(similarity_matrix, -self.batch_size) # bs
        positives = torch.cat([sim_ij, sim_ji], dim=0) # 2*bs
        nominator = torch.exp(positives / self.temperature) # 2*bs
        denominator = self.negatives_mask * torch.exp(similarity_matrix / self.temperature) # 2*bs, 2*bs
        loss_partial = -torch.log(nominator / torch.sum(denominator, dim=1)) # 2*bs
        loss = torch.sum(loss_partial) / (2 * self.batch_size)
        # loss = (2 * self.batch_size) / torch.sum(loss_partial)
        return loss
#===============================================================
# 上面是定义的计算两个特征的对比损失(余弦相似度)的函数(网上有这个对比损失函数，但是我找不到原网址了)，需要你定义在train函数之外，然后在train函数中调用 。
# 下面是完整的损失函数，你可以在train函数中直接这样使用。
#===============================================================
# 网络输出
output_fea, output_cls = net(data)
b_fea, c_fea = output_fea.shape
# ContrastiveLoss
output_fea_1 = output_fea[0:b_fea // 2, :]
output_fea_2 = output_fea[b_fea // 2:, :]
con_loss = ContrastiveLoss(b_fea // 2)
loss_view = con_loss.forward(output_fea_1, output_fea_2)
# CrossEntropyLoss
output_cls_1 = output_cls[0:b_fea // 2, :]
output_cls_2 = output_cls[b_fea // 2:, :]
loss_cls_1 = criterion(output_cls_1, target)
loss_cls_2 = criterion(output_cls_2, target)
# Multi-Loss
loss = loss_view + loss_cls_1 + loss_cls_2
#=====================================================

**Comment by ZJier** [comment]
整体训练和测试部分的代码，你可以参考：https://github.com/nshaud/DeepHyperX 或 [https://github.com/xiachangxue/DeepHyperX，你也可以很容易将这个工作嵌入进去。](https://github.com/xiachangxue/DeepHyperX%EF%BC%8C%E4%BD%A0%E4%B9%9F%E5%8F%AF%E4%BB%A5%E5%BE%88%E5%AE%B9%E6%98%93%E5%B0%86%E8%BF%99%E4%B8%AA%E5%B7%A5%E4%BD%9C%E5%B5%8C%E5%85%A5%E8%BF%9B%E5%8E%BB%E3%80%82)
对比学习部分的训练代码并不是独立的，它可以在上述两个链接中直接嵌入集成。下面是一个示例，你可以进行相关尝试：
首先，关于损失函数部分的代码如下：
`#===============================================================
class ContrastiveLoss(nn.Module):
    def init(self, batch_size, device='cuda', temperature=0.5):
        super().init()
        self.batch_size = batch_size
        self.register_buffer("temperature", torch.tensor(temperature).to(device)) # 超参数 温度
        self.register_buffer("negatives_mask", (~torch.eye(batch_size * 2, batch_size * 2, dtype=bool).to(device)).float()) # 主对角线为0，其余位置全为1的mask矩阵
    def forward(self, emb_i, emb_j): # emb_i, emb_j 是来自同一图像的两种不同的预处理方法得到
        z_i = F.normalize(emb_i, dim=1) # (bs, dim) ---> (bs, dim)
        z_j = F.normalize(emb_j, dim=1) # (bs, dim) ---> (bs, dim)
        representations = torch.cat([z_i, z_j], dim=0) # repre: (2bs, dim)
        similarity_matrix = F.cosine_similarity(representations.unsqueeze(1), representations.unsqueeze(0), dim=2) # simi_mat: (2bs, 2bs)
        sim_ij = torch.diag(similarity_matrix, self.batch_size) # bs
        sim_ji = torch.diag(similarity_matrix, -self.batch_size) # bs
        positives = torch.cat([sim_ij, sim_ji], dim=0) # 2bs
        nominator = torch.exp(positives / self.temperature) # 2bs
        denominator = self.negatives_mask * torch.exp(similarity_matrix / self.temperature) # 2bs, 2bs
        loss_partial = -torch.log(nominator / torch.sum(denominator, dim=1)) # 2bs
        loss = torch.sum(loss_partial) / (2 * self.batch_size)
        # loss = (2 * self.batch_size) / torch.sum(loss_partial)
        return loss
#===============================================================
上面是定义的计算两个特征的对比损失(余弦相似度)的函数(网上有这个对比损失函数，但是我找不到原网址了)，需要你定义在train函数之外，然后在train函数中调用 。
下面是完整的损失函数，你可以在train函数中直接这样使用。
#===============================================================

# 网络输出
output_fea, output_cls = net(data)
b_fea, c_fea = output_fea.shape

ContrastiveLoss
output_fea_1 = output_fea[0:b_fea // 2, :]
output_fea_2 = output_fea[b_fea // 2:, :]
con_loss = ContrastiveLoss(b_fea // 2)
loss_view = con_loss.forward(output_fea_1, output_fea_2)

CrossEntropyLoss
output_cls_1 = output_cls[0:b_fea // 2, :]
output_cls_2 = output_cls[b_fea // 2:, :]
loss_cls_1 = criterion(output_cls_1, target)
loss_cls_2 = criterion(output_cls_2, target)

Multi-Loss
loss = loss_view + loss_cls_1 + loss_cls_2
#=====================================================`

**Comment by ZJier** [comment]
1. https://github.com/nshaud/DeepHyperX 或 https://github.com/xiachangxue/DeepHyperX 有完整的训练和测试代码，你可以很容易将这个工作嵌入进去。
2. 关于对比学习的训练代码，它并不是独立的，你可以在上述完整代码链接中直接嵌入。这里是一个示例，你可以进行尝试：
首先，关于损失函数部分的代码如下：
######===============================================================
class ContrastiveLoss(nn.Module):
    def __init__(self, batch_size, device='cuda', temperature=0.5):
        super().__init__()
        self.batch_size = batch_size
        self.register_buffer("temperature", torch.tensor(temperature).to(device)) # 超参数 温度
        self.register_buffer("negatives_mask", (~torch.eye(batch_size * 2, batch_size * 2, dtype=bool).to(device)).float()) # 主对角线为0，其余位置全为1的mask矩阵
    def forward(self, emb_i, emb_j): # emb_i, emb_j 是来自同一图像的两种不同的预处理方法得到
        z_i = F.normalize(emb_i, dim=1) # (bs, dim) ---> (bs, dim)
        z_j = F.normalize(emb_j, dim=1) # (bs, dim) ---> (bs, dim)
        representations = torch.cat([z_i, z_j], dim=0) # repre: (2*bs, dim)
        similarity_matrix = F.cosine_similarity(representations.unsqueeze(1), representations.unsqueeze(0), dim=2) # simi_mat: (2*bs, 2*bs)
        sim_ij = torch.diag(similarity_matrix, self.batch_size) # bs
        sim_ji = torch.diag(similarity_matrix, -self.batch_size) # bs
        positives = torch.cat([sim_ij, sim_ji], dim=0) # 2*bs
        nominator = torch.exp(positives / self.temperature) # 2*bs
        denominator = self.negatives_mask * torch.exp(similarity_matrix / self.temperature) # 2*bs, 2*bs
        loss_partial = -torch.log(nominator / torch.sum(denominator, dim=1)) # 2*bs
        loss = torch.sum(loss_partial) / (2 * self.batch_size)
        # loss = (2 * self.batch_size) / torch.sum(loss_partial)
        return loss
######===============================================================
######上面是定义的计算两个特征的对比损失(余弦相似度)的函数(网上有这个对比损失函数，但是我找不到原网址了)，需要你定义在train函数之外，然后在train函数中调用 。
######下面是完整的损失函数，你可以在train函数中直接这样使用。
######===============================================================
######网络输出
output_fea, output_cls = net(data)
b_fea, c_fea = output_fea.shape
######ContrastiveLoss
output_fea_1 = output_fea[0:b_fea // 2, :]
output_fea_2 = output_fea[b_fea // 2:, :]
con_loss = ContrastiveLoss(b_fea // 2)
loss_view = con_loss.forward(output_fea_1, output_fea_2)
######CrossEntropyLoss
output_cls_1 = output_cls[0:b_fea // 2, :]
output_cls_2 = output_cls[b_fea // 2:, :]
loss_cls_1 = criterion(output_cls_1, target)
loss_cls_2 = criterion(output_cls_2, target)
######Multi-Loss
loss = loss_view + loss_cls_1 + loss_cls_2
######======================================================
